### PR TITLE
DP-118 Add API call scheduling capabilities to platform

### PR DESCRIPTION
**Pull request by yaroslavmo**
- Add failed scheduled tasks logs

### CODE REVIEW

File: src/Components/TaskScheduler.php [hunk 1]
```diff
@@ -17,24 +19,95 @@ class TaskScheduler
...
         if ($task->is_active) {
+
+            $dirPath = storage_path() . DIRECTORY_SEPARATOR . 'logs' . DIRECTORY_SEPARATOR . 'tmp';
+
+            if (!file_exists($dirPath)) mkdir($dirPath, 0777, true);
+
+            if ($logFile = self::createTaskLogFile($dirPath)) {
+                $logFilePath = stream_get_meta_data($logFile)['uri'];
+            } else {
+                $logFilePath = $dirPath . DIRECTORY_SEPARATOR . uniqid('schedule', true) . '.log';;
```
Comment (volodymyrpoli) [inline src/Components/TaskScheduler.php]: Double semicolon

File: src/Components/TaskScheduler.php [hunk 2]
```diff
@@ -17,24 +19,95 @@ class TaskScheduler
...
         $serviceName = \ServiceManager::getServiceById($task->service_id)->getName();
         $component = $task->component;
         $commandOptions = '--format=JSON --verb=' . $verb . ' --service=' . $serviceName . ' --resource=' . $component;
 
         // Use the scheduler to schedule the task at its desired frequency in minutes
         if ($task->is_active) {
+
+            $dirPath = storage_path() . DIRECTORY_SEPARATOR . 'logs' . DIRECTORY_SEPARATOR . 'tmp';
+
+            if (!file_exists($dirPath)) mkdir($dirPath, 0777, true);
```
Comment (volodymyrpoli) [inline src/Components/TaskScheduler.php]: I think it is a bad practice to create a directory with full access for all user

File: src/Components/TaskScheduler.php [hunk 3]
```diff
@@ -17,24 +19,95 @@ class TaskScheduler
...
-                ->onFailure(function () use ($task) {
+                ->sendOutputTo($logFilePath)
+                ->onSuccess(function () use ($task, $logFilePath) {
+                    if (TaskLog::whereTaskId($task->id)->exists()) {
+                        TaskLog::whereTaskId($task->id)->first()->delete();
+                    }
+                    if (file_exists($logFilePath)) unlink($logFilePath);
+                })
+                ->onFailure(function () use ($task, $logFilePath) {
+                    try {
```
Comment (volodymyrpoli) [inline src/Components/TaskScheduler.php]: Creating big anonymous methods is a bad practice. It is better to break down this method into at least two small

File: src/Components/TaskScheduler.php [hunk 4]
```diff
@@ -17,24 +19,95 @@ class TaskScheduler
...
+     * Get status code from error log
+     *
+     * @param string $fileName
+     * @return float|int|string
+     */
+    private static function getStatusCode($fileName)
+    {
+        $lines = file($fileName);
+        foreach ($lines as $lineNumber => $line) {
+            if (strpos($line, 'statusCode') !== false) {
```
Comment (volodymyrpoli) [inline src/Components/TaskScheduler.php]: Maybe it is better to use:
`if (strpos($line, 'statusCode') === true) {`

File: src/Components/TaskScheduler.php [hunk 5]
```diff
@@ -17,24 +19,95 @@ class TaskScheduler
...
+                return abs((int)filter_var($line, FILTER_SANITIZE_NUMBER_INT));
+            }
+        }
+        return '500';
```
Comment (volodymyrpoli) [inline src/Components/TaskScheduler.php]: In line 98, we return `int`, but here we return 'string'

File: src/Components/TaskScheduler.php [hunk 6]
```diff
@@ -17,24 +19,95 @@ class TaskScheduler
...
+    /**
+     * @param $dirPath
+     * @return array
+     */
+    private static function createTaskLogFile($dirPath)
+    {
+        putenv('TMPDIR='.$dirPath);
+        return tmpfile();
+    }
 }
```
Comment (volodymyrpoli) [inline src/Components/TaskScheduler.php]: Always add new line at end of file

File: src/Components/TaskScheduler.php (hunk 4 above)
Comment (yaroslavmo) [reply src/Components/TaskScheduler.php]: strpos returns FALSE if the needle was not found.
OR
strpos returns the position and not boolean. 

As I don't need the position and just want to know if the line exists !== false seems more appropriate.

File: src/Components/TaskScheduler.php (hunk 3 above)
Comment (yaroslavmo) [reply src/Components/TaskScheduler.php]: I don't know how I missed that. Refactored to several smaller functions

File: src/Components/TaskScheduler.php (hunk 2 above)
Comment (yaroslavmo) [reply src/Components/TaskScheduler.php]: I made it 0770. But this file will be deleted after file_get_contents function so I think it doesn't really matter.

File: src/Components/TaskScheduler.php (hunk 5 above)
Comment (yaroslavmo) [reply src/Components/TaskScheduler.php]: made it int

**Pull request merged**